### DISCUSSION TITLE
Please provide guidance on creating merkle tree for collection

### DISCUSSION THREAD

**Discussion by shea851** [ProjectOpenSea/seaport-js#77] [Q&A]
Hi,

Recently, criteria based offers began to show up from the offers API.

It seems that in order to fulfill an offer, an identifier must be supplied (token ID I'm assuming), as well as merkle proofs that this is a valid leaf in the tree. It also seems we are given the tree root.

I don't believe I can build a tree with just the root. I think the root is given so we can validate that the tree we built is a match.

I am not sure how to obtain these merkle proofs. **My guess is that I would need to build a tree. Can you please provide some guidance on this part?** Or am I totally missing something and there is an endpoint/contract I can call to get the proofs?

Thank you.

**Comment by shea851** [comment]
Just an update on where I'm at. I'm still stuck.

I've been testing with the 0N1 Force collection. It seems if you do a collection offer there are no filters (which makes this easy to test with), therefore I'm assuming a collection offer is just simply "collection-wide" and the merkle tree would consist of all token IDs when built.

So, I built a merkle tree with token IDs 1 - 7777 (inclusive), and upon generating proofs for one of the IDs, it seems I can get 4 of them to match, and the other 9 do not match. The root hash also does not match.  It's pretty cool that 4 do match, I think that means I'm on the right track, but not there yet.

At this point, I think I've exhausted ideas that make sense, and I've just been guessing at what could possibly be different. Why would the tree for a "collection-wide" offer have something other than all tokens in the collection? Perhaps it excludes tokens that are "locked" on OS?

**Comment by sunwrobert** [comment]
Hey @shea851, If we're not properly returning the proof that's on us. You shouldn't be expected to generate the proof yourself as we don't give you the list of valid identifiers either. This is something Opensea has to support returning from the REST endpoint.

**Reply by shea851** [reply]
@sunwrobert Thanks for the response. Is this something you can escalate to OpenSea to start to return from some REST endpoint? I believe I have reviewed them all and not found the way to retrieve either identifiers or proof for an offer order.

**Reply by sunwrobert** [reply]
Asking around

**Reply by warneracw21** [reply]
With this escalation, is it possible to simply identify which traits are being considered for by an offer? It seems impossible to understand how to map offers to specific traits with the current data returned by the REST endpoint

**Reply by shea851** [reply]
While working through this issue, I did come up with a theoretical way of doing it...
- Build a database with each trait and token IDs associated with it, for every collection
- Build merkle tree for each trait in each collection, save root hash
- Compare the root hash given by the order to find which trait it was for, then you can build tree and retrieve proofs

Problems with this approach
- In the test I did the root hash didn't seem to match OpenSea, and it made no sense why
- For new collections that are revealing, or collections that are still minting (MAYC, Clone-X, etc.) you would need to do this in a more dynamic way, which would require a lot of work to get performance where it should be
- There's no good way to get this data that I could find. You would have to ingest collection metadata yourself and build a pretty large database.

**Reply by warneracw21** [reply]
Hey @shea851 Shea, you want to work together on attacking this issue? Do you have discord? I have been working hard trying to implement a solution similar to yours. Pls DM me on the linked Twitter profile in my Github bio.

My discord is DiamondJim#5135

**Reply by shea851** [reply]
> Asking around

@sunwrobert Did you have any additional information on this? Thank you for your time.

**Reply by sunwrobert** [reply]
It's actively being worked on, next week should be reasonable

**Reply by warneracw21** [reply]
If possible, could seaport include a full example on how to fulfill an advanced order using the criteria proofs?

**Reply by AntHornerTech** [reply]
> It's actively being worked on, next week should be reasonable

Hey @sunwrobert are there any updates on this being provided to us from the REST API?

@warneracw21 @shea851 have you guys made any progress on your end?

**Reply by sunwrobert** [reply]
> > It's actively being worked on, next week should be reasonable
> 
> Hey @sunwrobert are there any updates on this being provided to us from the REST API?
> 
> @warneracw21 @shea851 have you guys made any progress on your end?

Hey, sorry things are a bit on hold now as we go through some restructuring.

**Reply by mustafa222lk** [reply]
is it updated now ? I still cant calculate the merkle proof. It returns an empty array

**Comment by Philippeomalley** [comment]
Hi, how exactly do you generate merkle tree for a range? say 1-10000? I'm not generating the correct root

**Comment by 0xMudded** [comment]
Is it not possible to let us know the range for generating the merkle tree?

**Comment by onurozansahin** [comment]
Hey all, are there any updates on this matter? What I've noticed is _sometimes_ the `identifierOrCriteria` returns a non-zero value which I believe are the proof that we look for. I believe the non-zero value is associated with the order being geared towards a certain trait.

My question is actually what to do when the value is returned as '0'. When I like to fulfil a particular order which is an offer for a collection valid for any trait ([hence the `identifierOrCriteria` of '0'](https://docs.opensea.io/v2.0/reference/seaport-overview#order)), the SDK returns the error, "You must supply the appropriate criterias for criteria based items".

What to do in these cases? Is it required to supply a proof ourselves? How does that relate to the signature provided with the order? Can a consideration be extended not only for tipping, but also for modifying the proof? Many head scratching questions are floating...

**Reply by nfts4f4fff** [reply]
You pass in an empty array basically

**Comment by Alotrobo** [comment]
No news about the implementation in the endpoint ?

**Comment by jrico207** [comment]
Why not just provide the information on how to calculate the tree instead of providing an endpoint?

**Reply by nazariyv** [reply]
let me know if this is helpful: https://medium.com/@nazariyv/merkle-tree-for-nft-offers-on-marketplaces-1d24e8c61ed4

**Reply by mustafa222lk** [reply]
doesnt help. I just get an empty array every time

**Comment by mustafa222lk** [comment]
Anyone here was able to accept a collection based offer using this libary ?

**Comment by Ha2ryZhang** [comment]
anything update?

**Comment by mustafa222lk** [comment]
where is the support wtf

**Reply by mustafa222lk** [reply]
@sunwrobert why are you not helping any of us ? This doesnt look good. You are literally ignoring everything and just gave attention to the older ones 7-8months ago

**Comment by nfts4f4fff** [comment]
Same issue for me. After 7 months there is still no update. I can't believe this

**Comment by JrCx7scC** [comment]


**Reply by mustafa222lk** [reply]
the seaport api endpoint for the trair and collection based offers are literally useless without providing us this merkle proof information

**Comment by Alotrobo** [comment]
Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)

**Reply by mustafa222lk** [reply]
we need to know how the merkle tree was built in order to provide a merkle proof accordingly. May I know how you did it. This would help me and other people here a lot

**Reply by nfts4f4fff** [reply]
> Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)

I don't think that is true. Many people here are trying and it doesn't work like that

**Reply by nfts4f4fff** [reply]
If it is true. Please provide the solution to us

**Reply by JrCx7scC** [reply]
the code in the unit test is successful . check and try unit tests

**Reply by mustafa222lk** [reply]
could you provide a link ? I sadly couldnt find anything called with unit

**Reply by mustafa222lk** [reply]
I tried the examples in the test folder but sadly it returns an empty array except the needed proof. Maybe a code block of the solution would be amazing

**Reply by mustafa222lk** [reply]
`proof: new MerkleTree([nftId]).getProof(nftId),`
This doesnt work. I get a empty array

**Reply by nfts4f4fff** [reply]
> the code in the unit test is successful . check and try unit tests

Any code share ?

**Reply by mustafa222lk** [reply]
> the code in the unit test is successful . check and try unit tests

please >?

**Reply by Alotrobo** [reply]
> > Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)
> 
> I don't think that is true. Many people here are trying and it doesn't work like that

However it is the case, the criteria is functional as I can accept collections offers, the code is the same as theirs.

**Reply by mustafa222lk** [reply]
> > > Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)
> > 
> > 
> > I don't think that is true. Many people here are trying and it doesn't work like that
> 
> However it is the case, the criteria is functional as I can accept collections offers, the code is the same as theirs.

could you please share the functional code for us all

**Reply by mustafa222lk** [reply]
> > > Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)
> > 
> > 
> > I don't think that is true. Many people here are trying and it doesn't work like that
> 
> However it is the case, the criteria is functional as I can accept collections offers, the code is the same as theirs.

could you show me which code works please ? Im really waiting for it for days now as the creators dont care at all helping the community

**Reply by nfts4f4fff** [reply]
> the code in the unit test is successful . check and try unit tests

Please help ?

**Reply by nfts4f4fff** [reply]
> > > Everything is in their github, you just have to set up the merkle using their criteria code, and everything will work (I did it)
> > 
> > I don't think that is true. Many people here are trying and it doesn't work like that
> 
> However it is the case, the criteria is functional as I can accept collections offers, the code is the same as theirs.

Week passed and still nothing
